### DISCUSSION TITLE
Git minimum version check

### DISCUSSION THREAD

**Discussion by shorebvba** [jwodder/versioningit#10] [Ideas]
It might be a good idea to specify the minimum Git version, as currently you get misleading errors when using 1.8 for example, which does not have the -C option apparently.
Checking version before using -C in versioningit code would be ideal, but in the readme would also suffice i guess.

If this is a pebcac, i apologize, if not i'm willing to contribute.

**Comment by jwodder** [comment]
So here are all of the Git features that `versioningit` uses, in no particular order, (not counting `$Format:%(describe)$`, which it doesn't use directly) and the Git versions that introduced them:

* `git -C <dir>` option — added in Git 1.8.5
* `git --no-pager` option — added in Git 1.5.3
* `git rev-parse --is-inside-work-tree` — added in Git 1.5.3
* `git ls-files --error-unmatch .` — added in (or at least existed in) Git 1.3.0
* `git describe`:
    * The output only started including the distance in Git 1.5.0.
    * `--tags` option — existed in Git 1.1.0
    * `--long` option — added in Git 1.5.5
    * `--dirty` option — added in Git 1.6.6
    * `--always` option — added in (or at least existed in) Git 1.5.6
    * `--match` option — added in Git 1.5.5; support for specifying multiples times added in Git 2.13.0
    * `--exclude` option — added in Git 2.13.0
* `git symbolic-ref --short -q HEAD` — fully implemented in Git 1.7.10
* `git show`:
    * `-s` option — existed prior to Git 1.0
    * `--format` option — added in Git 1.6.3
    * `%H`, `%n`, `%at`, and `%ct` format placeholders — all added in Git 1.5.1
* `git rev-list --count HEAD` — added in Git 1.7.2

Aside from `--match` and `--exclude` (which are only used if the user requests them), the most recent feature here is the `-C` option, which is easy to compensate for.  The next most recent is `git symbolic-ref --short -q HEAD` (used to determine the branch name), which I am not aware of a decent older alternative for.

So, assuming I can spin up a Docker container with Git 1.8.0 and a recent Python and get `versioningit`'s non-match/exclude tests to pass in it, I can then make a new release that doesn't use the `-C` option, and I will document 1.8.0 as the minimum required Git version in the README.

I'd rather not introduce an explicit Git version check into the code; [there are already concerns about the Git calls taking up too much time](https://github.com/jwodder/versioningit/pull/7), and I expect there to be very few users using an over-ten-years-old version of Git (especially considering that `versioningit` currently doesn't support any Python versions more than six years old).

**Reply by shorebvba** [reply]
Wow, for me this post alone helps a lot. Thank you!
We are currently stuck with 1.8.3.1 on a CentOS server. So not the end of the world (yet, hehe).
I didn't realize the speed hit on git calls.

Great work! Thank you.

**Reply by jwodder** [reply]
Update: `versioningit` v0.3.3 is now out with support for Git 1.8.0.

**Reply by shorebvba** [reply]
wow, amazing speed John, thank you so much
will try this week

**Reply by shorebvba** [reply]
fixed our issues, working on upgrading git anyway

A+++ service, would recommend :)
